enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,1098 @@
+reduce the likelihood of datacenter level faults degrading the control plane.
+To this end, the failure domains for the control plane will be set on the `ControlPlaneMachineSet` spec directly.
+
+When creating Machines, the `ControlPlaneMachineSet` controller will balance the Machines across these failure domains
+by injecting the desired failure domain for the new Machine into the provider spec based on the platform specific
+failure domain field.
+
+##### Failures domains
+
+The `FailureDomains` field is expected to be populated by the user/installer based on the topology of the Control Plane
+Machines. For example, we expect on AWS that this will contain a list of availability zones within a single region.
+Note, we are explicitly not expecting users to add different regions to the the `FailureDomains` as we do not support
+running OpenShift across multiple regions.
+
+Note that the `FailureDomains` field is only supported on certain platforms, currently; AWS, Azure, GCP and OpenStack;
+other platforms may be supported in the future.
+
+The users will be allowed to override a small amount of configuration for the `providerSpec` based on the configuration
+required to spread Machines across different failure domains.
+For example, on AWS, both the `availabilityZone` and `subnets` differ depending on which failure domain is configured,
+on other platforms, eg Azure, GCP or OpenStack, only one field is required to vary the failure domain, in which case,
+this is all that will be allowed.
+
+The overrides will be injected into the given `providerSpec` before creating the Machines as part of the balancing logic
+within the `ControlPlaneMachineSet` operator.
+
+As an example, a user on AWS may set their `FailureDomains` as:
+
+```yaml
+failureDomains:
+  aws:
+  - placement:
+      availabilityZone: us-east-1a
+    subnet:
+      filters:
+      - name: "tag:Name"
+        values:
+        - "my-cluster-subnet-1a"
+  - placement:
+      availabilityZone: us-east-1b
+      subnet:
+        filters:
+        - name: "tag:Name"
+          values:
+          - "my-cluster-subnet-1b"
+```
+
+A user on Azure may set their `FailureDomains` as:
+
+```yaml
+failureDomains:
+  azure:
+  - zone: us-central-1
+  - zone: us-central-2
+```
+
+###### Failure Domains on vSphere
+
+As of OpenShift 4.10, there is no concept of a zone within vSphere and as such we would expect that the `FailureDomains`
+would be omitted for vSphere.
+
+However, [there is future work](https://github.com/openshift/enhancements/pull/918) to include zone support for vSphere
+within 4.11. Once this is implemented, the Machine API provider for vSphere will understand the concept of zones,
+defined within the `Infrastructure` resource.
+The `ControlPlaneMachineSet` will then be able to balance Machines across the different zones by setting the new zone
+information within the provider specs as it does on other platforms.
+
+#### Ensuring Machines match the desired state
```

Comment:
@JoelSpeed Is there any reason why a control plane should not perform veritical scaling while the control plane is upgrading? My thoughts are this should be ok, just wanted to raise this question. Apologies if its been raised elsewhere.

cc @jewzaam

Reply:
I think this was previously discussed and we concluded that because the upgrade procedure itself is disruptive to the control plane, mixing vertical scaling into this could cause significant extra disruption and elongate both the process of upgrade and the rollout.

As these are both high risk scenarios, I think we concluded that it would be better to have them be mutually exclusive, though I can't find the previous conversation now

Reply:
Thanks @JoelSpeed I think i remember discussing elsewhere as well :man_shrugging:  - is there a webhook or another implimentation to prevent triggering veritical scaling during upgrades?

Reply:
This will be prevented within the controller itself, ie, users will be able to make the change, but the operator will mark itself as unable to rollout due to the pending update